worker.service_sched: processing eval panicked scheduler - please report this as a bug

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Nomad issue
tracker reserved for bug reports and feature requests. For general usage
questions, please use: https://discuss.hashicorp.com/c/nomad/28

-->

### Nomad version
```
Nomad v2.0.4
BuildDate 2026-07-07T17:20:59Z
Revision 5b83b133998a1f514beb81019930fb673b5ed669
```

### Operating system and Environment details
```
lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 24.04.4 LTS
Release:	24.04
Codename:	noble

uname -r
6.8.0-136-generic
```

### Issue
nomad job run gets stuck and this crash occurs in the log:
```
Jul 24 15:29:08 server-app nomad[14739]:     2026-07-24T15:29:08.763Z [ERROR] worker.service_sched: processing eval panicked scheduler - please report this as a bug!: eval_id=e01bfc17-c869-714f-6157-0fcc0e3a4180 job_id=redacted namespace=default worker_id=a24ec0a2-c88f-03c1-2bb8-0c093d339043 eval_id=e01bfc17-c869-714f-6157-0fcc0e3a4180 error="runtime error: invalid memory address or nil pointer dereference"
Jul 24 15:29:08 server-app nomad[14739]:   stack_trace=
Jul 24 15:29:08 server-app nomad[14739]:   | goroutine 63 [running]:
Jul 24 15:29:08 server-app nomad[14739]:   | runtime/debug.Stack()
Jul 24 15:29:08 server-app nomad[14739]:   | \truntime/debug/stack.go:26 +0x5e
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler.(*GenericScheduler).Process.func1()
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/generic_sched.go:108 +0x58
Jul 24 15:29:08 server-app nomad[14739]:   | panic({0x3652ec0?, 0x6b69fd0?})
Jul 24 15:29:08 server-app nomad[14739]:   | \truntime/panic.go:860 +0x13a
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*HostVolumeChecker).hostVolumeIsAvailable(0x116a953735e0, 0x116a941bc900, {0x116a940dd5a0?, 0xcc36e484c52f100?}, {0x116a958b43f0?, 0x0?}, 0x0, {0x116a959fb380, 0x5, 0x5})
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/feasible.go:476 +0x52c
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*HostVolumeChecker).nodeIsFeasibleForRequest(0x116a953735e0, 0x116a940a9200?, 0x116a94196000, {0x116a959fb380, 0x5, 0x5})
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/feasible.go:361 +0x171
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*HostVolumeChecker).hasVolumes(0x116a953735e0, 0x116a953b84e0)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/feasible.go:333 +0xc9
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*HostVolumeChecker).Feasible(0x116a953735e0, 0x116a953b84e0)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/feasible.go:313 +0x25
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*FeasibilityWrapper).available(0x378b460?, 0x116a953b84e0)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/feasible.go:1519 +0x6b
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*FeasibilityWrapper).Next(0x116a9553f080)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/feasible.go:1501 +0x79e
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*DistinctHostsIterator).Next(0x116a94520a40)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/feasible.go:861 +0x4b
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*DistinctPropertyIterator).Next(0x116a94de42a0)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/feasible.go:976 +0x2c
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*FeasibleRankIterator).Next(0xa?)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/rank.go:105 +0x1c
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*BinPackIterator).Next(0x116a95373650)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/rank.go:215 +0x62
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*JobAntiAffinityIterator).Next(0x116a9536e4b0)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/rank.go:835 +0x68
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*NodeReschedulingPenaltyIterator).Next(0x116a95ab3290)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/rank.go:896 +0x25
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*NodeAffinityIterator).Next(0x116a9536e500)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/rank.go:968 +0x2d
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*SpreadIterator).Next(0x116a95373ab0)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/spread.go:131 +0x30
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*PreemptionScoringIterator).Next(0x116a949c4440)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/rank.go:1061 +0x25
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*ScoreNormalizationIterator).Next(0x116a949c4460)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/rank.go:1025 +0x25
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*LimitIterator).nextOption(0x116a94de4300)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/select.go:63 +0x22
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*LimitIterator).Next(0x116a94de4300)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/select.go:42 +0x26
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*MaxScoreIterator).Next(0x116a95ab3950)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/select.go:105 +0x3b
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/feasible.(*GenericStack).Select(0x116a94aacc00, 0x116a94744ea0, 0x116a9536f360)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/feasible/stack.go:201 +0xfdf
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler.(*GenericScheduler).computeJobAllocs.genericAllocUpdateFn.func1(0x116a94b58400, 0x116a953f1448, 0x116a94744ea0)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/util.go:871 +0x3a3
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/reconciler.(*AllocReconciler).computeUpdates(0x116a94649ce0, 0x116a9570a870, 0x116a94744ea0, 0x116a94649e40)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/reconciler/reconcile_cluster.go:1391 +0x153
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/reconciler.(*AllocReconciler).computeGroup(0x116a94649ce0, {0x116a94266668, 0x8}, 0x116a9570a570)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/reconciler/reconcile_cluster.go:496 +0x141a
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/reconciler.(*AllocReconciler).computeDeploymentComplete(0x116a94649ce0, 0x116a94649d90, 0x116a958b48a6?)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/reconciler/reconcile_cluster.go:368 +0x85
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler/reconciler.(*AllocReconciler).Compute(0x116a94649ce0)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/reconciler/reconcile_cluster.go:305 +0x33d
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler.(*GenericScheduler).computeJobAllocs(0x116a94649ad0)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/generic_sched.go:359 +0x45a
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler.(*GenericScheduler).process(0x116a94649ad0)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/generic_sched.go:245 +0x60a
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler.retryMax(0x5, 0x116a94cdfd88, 0x116a94cdfcf8)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/util.go:70 +0x49
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/scheduler.(*GenericScheduler).Process(0x116a94649ad0, 0x116a94c38000)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/scheduler/generic_sched.go:143 +0x53b
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/nomad.(*Worker).invokeScheduler(0x116a94d41420, 0x116a95ab2690, 0x116a94c38000, {0x116a94d6e360, 0x24})
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/nomad/worker.go:635 +0x2ee
Jul 24 15:29:08 server-app nomad[14739]:   | github.com/hashicorp/nomad/nomad.(*Worker).run(0x116a94d41420, 0x12a05f200)
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/nomad/worker.go:464 +0x54f
Jul 24 15:29:08 server-app nomad[14739]:   | created by github.com/hashicorp/nomad/nomad.(*Worker).Start in goroutine 1
Jul 24 15:29:08 server-app nomad[14739]:   | \tgithub.com/hashicorp/nomad/nomad/worker.go:163 +0x59
```

--- a/.changelog/28301.txt
+++ b/.changelog/28301.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+scheduler: Fixed a bug where the scheduler could panic with a nil pointer dereference when checking host volume feasibility for an allocation whose job had been purged
+```

--- a/scheduler/feasible/feasible.go
+++ b/scheduler/feasible/feasible.go
@@ -473,7 +473,18 @@ func (h *HostVolumeChecker) hostVolumeIsAvailable(
 			if err != nil {
 				return false
 			}
-			for _, req := range job.LookupTaskGroup(alloc.TaskGroup).Volumes {
+			// The job may have been purged or garbage collected (JobByID
+			// returns nil, nil), or the alloc's task group may no longer exist.
+			// We cannot prove this allocation is not using the volume, so treat
+			// it as unavailable rather than dereference the nil job.
+			if job == nil {
+				return false
+			}
+			tg := job.LookupTaskGroup(alloc.TaskGroup)
+			if tg == nil {
+				return false
+			}
+			for _, req := range tg.Volumes {
 				if vol.MatchesRequestSource(req, alloc) {
 					if !req.ReadOnly {
 						return false

--- a/scheduler/feasible/feasible_test.go
+++ b/scheduler/feasible/feasible_test.go
@@ -902,6 +902,36 @@ func TestDynamicHostVolumeIsAvailable(t *testing.T) {
 
 }
 
+func TestDynamicHostVolumeIsAvailable_PurgedJob(t *testing.T) {
+	ci.Parallel(t)
+
+	_, ctx := MockContext(t)
+	checker := NewHostVolumeChecker(ctx)
+
+	vol := &structs.HostVolume{
+		Name:  "example",
+		State: structs.HostVolumeStateReady,
+		RequestedCapabilities: []*structs.HostVolumeCapability{{
+			AttachmentMode: structs.HostVolumeAttachmentModeFilesystem,
+			AccessMode:     structs.HostVolumeAccessModeSingleNodeSingleWriter,
+		}},
+	}
+
+	// A proposed alloc whose job has been purged or garbage collected: JobByID
+	// returns nil, nil. The checker cannot prove the alloc is not using the
+	// volume, so it must report the volume unavailable rather than dereference
+	// the nil job (#28301).
+	orphan := mock.Alloc()
+
+	must.False(t, checker.hostVolumeIsAvailable(
+		vol,
+		structs.HostVolumeAccessModeSingleNodeSingleWriter,
+		structs.HostVolumeAttachmentModeFilesystem,
+		false,
+		[]*structs.Allocation{orphan},
+	))
+}
+
 func TestCSIVolumeChecker(t *testing.T) {
 	ci.Parallel(t)
 	state, ctx := MockContext(t)

